### DISCUSSION TITLE
Guidance on Adding Support for BPE Tokenizer in F5-TTS

### DISCUSSION THREAD

**Discussion by Alykasym** [SWivid/F5-TTS#748] [Q&A] [answered]
Hi,

I’ve been working with the F5-TTS model, and I’m interested in integrating a Byte Pair Encoding (BPE) tokenizer instead of the character-level tokenizer currently used. I’m familiar with how BPE tokenization works and the general steps for implementing it, but I’m unsure of the specific files and methods in the F5-TTS codebase that need to be modified to support this change.

Could you kindly guide me on:

1. Which files should be modified to integrate a BPE tokenizer into the pipeline?
2. Which methods/functions in the code deal with text tokenization, and how can I adapt them to work with BPE?
3. Any particular parts of the training and inference code that need to be adjusted to accommodate the new tokenizer? (for example, text preprocessing, data loading, model inputs, etc.)
4. If there’s a preferred way of adding custom tokenizers or if there are any specific considerations I should be aware of when implementing this change?
5. I would greatly appreciate any tips, documentation, or pointers on where to start.

Thanks in advance!

**Comment by SWivid** [comment, marked as answer]
core codes with model here https://github.com/SWivid/F5-TTS/tree/main/src/f5_tts/model

basically change the text format is fine https://github.com/SWivid/F5-TTS/blob/c2cf31e0c56c0840cb4fc9d6e4242f0d379fc50c/src/f5_tts/model/cfm.py#L214
https://github.com/SWivid/F5-TTS/blob/c2cf31e0c56c0840cb4fc9d6e4242f0d379fc50c/src/f5_tts/model/cfm.py#L227-L233

`text` is currently input like: [['h', 'o', 'w', ' ', 'a', 'r', 'e', ' ', 'y', 'o', 'u', '?',], ['i', '\'', 'm', ' ', 'f', 'i', 'n', 'e', '.',]]
you could just change to bpe.

inference process is the same, you could make a func e.g. char_to_bpe(), and pass the result in as `text` value

**Reply by Alykasym** [reply]
Thank you so much. I will post updates once I have results

**Comment by abhisirka2001** [comment]
What was the reason you shifted to bpe tokenizer? just curios and it might help me  because i am also facing issues in vocab learning .
@Alykasym

**Reply by Alykasym** [reply]
Main reason was that I had a small dataset and I was trying to teach a new language with it.

When I finetuned using char tokenizer, it did produce pretty good results, but sometimes it pronounced some vowels longer or shorter than it should be.

Basically, char tokenizer needs more data for the model to learn the different scenario pronounciations.

So I thought maybe BPE tokenizer will help, as it tokenizes chunks of letters in pairs. So it will hold more context of scenarios for the model to learn.

BPE did indeed reduce the frequency of wrong pronouncations but they were still there. Basically, I still need larger dataset.

**Comment by abhisirka2001** [comment]
Ohh gotcha
I am also facing issues related to pronunciation but yes i have increased the data volume so it might get resolved. Lets see.. thanks for help

**Comment by SachinTelecmi** [comment]
@Alykasym can you share your settings.json file?? i alos tried to train f5-tts with open source datya espcially for hindi but my results are not that great as its should be espcially the pronunciation, whats your data size and how many epochs you trained!! i was training with 4090 i tried to train it for a week still the results were not that upto mark !!

**Reply by Alykasym** [reply]
yeah sure, here's my settings.json content:

```
{
    "exp_name": "F5TTS_v1_Base",
    "learning_rate": 1e-05,
    "batch_size_per_gpu": 2,
    "batch_size_type": "sample",
    "max_samples": 64,
    "grad_accumulation_steps": 1,
    "max_grad_norm": 1,
    "epochs": 1905,
    "num_warmup_updates": 692,
    "save_per_updates": 27710,
    "keep_last_n_checkpoints": -1,
    "last_per_updates": 6928,
    "finetune": true,
    "file_checkpoint_train": "",
    "tokenizer_type": "char",
    "tokenizer_file": "",
    "mixed_precision": "none",
    "logger": "tensorboard",
    "bnb_optimizer": false
}

```


btw, I just trained for 100 epochs and this specific run is full FP32 precision. But I tried with BF16 and FP8, and all of them gave similar results, not much difference.

My Dataset size was 35 hours - very accurate hand written transcriptions and manually split studio quality audios. Multispeaker, around 15 speakers.

P.s:
There's an issue with `"convert_char_to_pinyin"` in utils.py. Basically, when you choose char tokenizer, the code somehow initiates `"convert_char_to_pinyin"` and it ends up adding extra spaces between characters. To fix it, you can just remove the content of `convert_char_to_pinyin` and make it return content as it is, or comment it out wherever it is called.

**Reply by Alykasym** [reply]
One of the things I can suggest is to use "sample" batch type.
After 100,000 steps, it should already start producing good results. If it doesn't, then something wrong with the dataset or the code, and no need to train further, it won't get better.

**Reply by dpp-user** [reply]
are you changing learning throughout the training? why are you using sample and not frame?

**Reply by dpp-user** [reply]
also i noticed in the tensor board that my loss starts at 0.7 but ends in 0.6 like its not improving but the samples generation does improving. so im confused with all of that new language training.

**Reply by Alykasym** [reply]
- Project script uses linear LR by default, so yeah, it slowly decreases the LR longer you train.
- When I tried Sample batch type, I didn't get any repeated words or skipped words, it was pronouncing all the words without skipping and repeating.
- 0.7-0.6 loss is normal range. The loss shouldn't decrease much if you train new language, it should go sideways. Just monitor whether there is no big spikes, like loss suddenly goes above 1.0 loss or below 0.4 loss. Like sudden shift of loss means there might be something wrong.

**Reply by dpp-user** [reply]
have you tried other learning rate other than 1e-05? how can we know what learning rates work well in this trainer?  i know from other ai training i do i usually start high in learning rate then lower it when training progress slows.

another question i have, once iv finished the fine tuning for the new language can i use it as a starting point and improve it from time to time? like adding more voices diversity, more data or even start training a specific voice in that language on top of this model and not the base f5-tts one?

**Reply by Alykasym** [reply]
- I tried different LR, but for my dataset size 1e-05 works well. It depends on your dataset size and whether you want to train longer. If you have bigger dataset and you want to train for few hundred epochs, then you can lower the LR further.
- Yes, you can use the finetuned model to finetune further with new data. One of the example scenarios would be:
Let's say you have 200 hours of non verified (not clean/not high quality) dataset. You finetune the model with that dataset, and after that you get like 50 hours of super high quality accurate dataset and continue the finetuning with it. Basically, you will be doing Base model unsupervised training and supervised training afterwards to get high quality dataset while maintaining the knowledge of the language.

But when you do the second finetuning, don't run it long. 100,000-150,000 steps should be enough. Otherwise it will forget the details of your first dataset and just remember the second dataset.

**Reply by dpp-user** [reply]
thank you, but how do i use the trained model as a base for the next? somehow when i enter it as path to pretrained model it not using it. cause when i generate at the beginning of the training it still generate gibrish or japanese and it suppose to already know the new language just fine tune on a specific voice

**Reply by Alykasym** [reply]
Not sure, it worked for me. You can open an issue about it with the details like Source code version, and steps to reproduce. Maybe if other people encountered it, they might be able to help.

**Reply by campar** [reply]
@Alykasym I've noticed that my dataset contains duplicate entries, the same text with two different audio files (not identical, but very similar). Could this affect the quality of training?

**Reply by Alykasym** [reply]
> @Alykasym I've noticed that my dataset contains duplicate entries, the same text with two different audio files (not identical, but very similar). Could this affect the quality of training?

It shouldn't affect as long as it is different person's voice. But if it is same person's voice and same text transcription, then it might affect badly.
